Sources/Utility/ProcessSet.swift

```diff
@@ -0,0 +1,127 @@
+/*
+ This source file is part of the Swift.org open source project
+ 
+ Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+ 
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Dispatch
+import Foundation
+import Basic
+
+typealias Thread = Basic.Thread
+
+public enum ProcessSetError: Swift.Error {
+    /// The process group was cancelled and doesn't allow adding more processes.
+    case cancelled
+}
+
+/// A process set is a small warpper for collection of processes.
+/// 
+/// This class is thread safe.
+public final class ProcessSet {
+
+    /// Array to hold the processes.
+    private var processes: Set<Process> = []
+
+    /// Queue to mutate internal states of the process group.
+    private let serialQueue = DispatchQueue(label: "org.swift.swiftpm.process-set")
+
+    /// If the process group was asked to cancel all active processes.
+    private var cancelled = false
+
+    /// The timeout (in seconds) after which the processes should be killed if they don't respond to SIGINT.
+    public let killTimeout: Double
+
+    /// Condition to block kill thread until timout.
+    private var killingCondition = Condition()
+
+    /// Boolean predicate for killing condition.
+    private var shouldKill = false
+
+    /// Create a process set.
+    public init(killTimeout: Double = 5) {
+        self.killTimeout = killTimeout
+    }
+
+    /// Add a process to the process set. This method will throw if the process set is already terminated.
+    ///
+    /// Call remove() method to remove the process from set once it has terminated.
+    ///
+    /// - Parameters:
+    ///   - process: The process to add.
+    /// - Throws: ProcessGroupError
+    public func add(_ process: Process) throws {
+        return try serialQueue.sync {
+            guard !cancelled else {
+                throw ProcessSetError.cancelled
+            }
+            self.processes.insert(process)
+        }
+    }
+
+    /// Remove the process from process set.
+    public func remove(_ process: Process) {
+        serialQueue.sync {
+            self.processes.remove(process) 
+            // If all processes terminated and we're cancelled, unblock the killing thread.
+            if cancelled && self.processes.isEmpty {
+                killingCondition.whileLocked {
+                    shouldKill = true
+                    killingCondition.signal()
+                }
+            }
+        }
+    }
+
+    /// Terminate all the processes. This method blocks until all processes in the set are terminated.
+    ///
+    /// A process set cannot be used once it has been asked to terminate.
+    public func terminate() {
+        serialQueue.sync {
+            // We're done if already cancelled.
```

Comment:
This is confusing, the `return` statements here don't return from `terminate` so what are they really a fast path on, and why would we need a fast path anyway?